tutorials/preprocessing/plot_40_artifact_correction_ica.py

```diff
@@ -1,304 +1,493 @@
+raw.plot(order=artifact_picks, n_channels=len(artifact_picks))
 
 ###############################################################################
-# Define the ICA object instance
-ica = ICA(n_components=n_components, method=method, random_state=random_state)
-print(ica)
+# We can get a better view of the EOG artifacts using
+# :func:`~mne.preprocessing.create_eog_epochs` like we did in the
+# :ref:`tut-artifact-overview` tutorial:
 
-###############################################################################
-# we avoid fitting ICA on crazy environmental artifacts that would
-# dominate the variance and decomposition
-reject = dict(mag=5e-12, grad=4000e-13)
-ica.fit(raw, picks='meg', decim=decim, reject=reject)
-print(ica)
+eog_evoked = create_eog_epochs(raw).average()
+eog_evoked.plot_joint()
 
 ###############################################################################
-# Plot ICA components
-ica.plot_components()  # can you spot some potential bad guys?
+# Now we'll do the same for the ECG artifacts, using
+# :func:`~mne.preprocessing.create_ecg_epochs`:
 
+ecg_evoked = create_ecg_epochs(raw).average()
+ecg_evoked.plot_joint()
 
 ###############################################################################
-# Component properties
-# --------------------
-#
-# Let's take a closer look at properties of first three independent components.
+# Before we run the ICA, an important step is filtering the data to remove
+# low-frequency drifts, which can negatively affect the quality of the ICA fit.
+# A 1 Hz cutoff frequency is recommended; we'll keep a copy of the unfiltered
+# :class:`~mne.io.Raw` object around so we can apply the ICA solution to it
+# later.
 
-# first, component 0:
-ica.plot_properties(raw, picks=0)
+filt_raw = raw.copy()
+filt_raw.load_data().filter(l_freq=1., h_freq=None)
 
 ###############################################################################
-# we can see that the data were filtered so the spectrum plot is not
-# very informative, let's change that:
-ica.plot_properties(raw, picks=0, psd_args={'fmax': 35.})
+# .. sidebar:: ICA ignores the time domain
+#
+#     ICA finds patterns across channels, but ignores the time domain. This
+#     makes it just as valid to compute ICA on discontinuous
+#     :class:`~mne.Epochs` objects as on continuous :class:`~mne.io.Raw`
+#     objects, or to only use every Nth sample by passing the ``decim``
+#     parameter to the :meth:`~mne.preprocessing.ICA.fit` method.
+#
+# Now we're ready to set up and fit the ICA. Since we know (from observing our
+# raw data) that the EOG and ECG artifacts are fairly strong, we would expect
+# those artifacts to be captured in the first few dimensions of the PCA
+# decomposition that happens before the ICA. Therefore, we probably don't need
+# a huge number of components to do a good job of isolating our artifacts. As a
+# first guess, we'll run ICA with only the first 15 PCA components — a fairly
```

Comment:
I tend to default to 20 because that is exactly one full figure.

Reply:
I picked 15 for similar reasons: because it yielded an integer number of complete rows in the component plot.  But the tutorial already takes 62 seconds to build on my laptop, so I'm not going to increase it here.